[MPDX-8542] Show support goal history in donations graph

canac opened the pull request:
## Description

Show the historical support goal in the donations graph.

## Testing

Testing is a little bit complex because the donations graph shows up in multiple places and each place loads the data independently (it isn't loaded by the graph component itself), so we need to test each place the donations graph is rendered as a HI user and a non-HI user.

* User without HI data
  * Go to the donations graph on the dashboard for a user without HI data. Check that the monthly goal is a straight line and is the value set in preferences.
  * Go to the donations graph on the donations report page for a user without HI data. Check that the monthly goal is a straight line and is the value set in preferences.
  * Go to the donations graph on the coaching page (viewing the user's own data) for a user without HI data. Check that the monthly goal is a straight line and is the value set in preferences.
  * Make the user coach Pedro and go to Pedro's coaching page (viewing Pedro's data as a coach). Check that the monthly goal line is the HI values for 2024 and is the current monthly goal preference for 2025.
* User with HI data (Pedro)
  * Go to the donations graph on the dashboard for a user without HI data. Check that the monthly goal line is the HI values for 2024 and is the current monthly goal preference for 2025.
  * Go to the donations graph on the donations report page for a user without HI data. Check that the monthly goal line is the HI values for 2024 and is the current monthly goal preference for 2025.
  * Go to the donations graph on the coaching page (viewing the user's own data) for a user without HI data. Check that the monthly goal line is the HI values for 2024 and is the current monthly goal preference for 2025.
  * Make the user coach a user without HI data and go to that user's coaching page (viewing the non-UI user as Pedro). Check that the monthly goal is a straight line and is the value set in preferences.

[MPDX-8542](https://jira.cru.org/browse/MPDX-8542)

## Checklist:

- [x] I have given my PR a title with the format "MPDX-(JIRA#) (summary sentence max 80 chars)"
- [x] I have applied the appropriate labels. (_Add the label "On Staging" to get the branch automatically merged into staging._)
- [x] I have requested a review from another person on the project

dr-bizz commented on src/components/Dashboard/DonationHistories/DonationHistories.tsx:
```diff
@@ -242,15 +238,23 @@ const DonationHistories = ({
                       <Legend />
                       <CartesianGrid vertical={false} />
-                      {goal && (
+                      {data?.healthIndicatorData?.length === 0 ? (
```
Did you make it `=== 0` to make it more readable that it' has no length?
```suggestion
                      {!data?.healthIndicatorData?.length ? (
```

dr-bizz reviewed (commented):
This is looking great! Good work. I have a few questions.

1. The coaching graph looked different than the other graph's, the goal was just a straight line rather being a dot with a line.
 
 Coaching page:
<img width="1728" alt="Screenshot 2025-02-14 at 10 53 49 AM" src="https://github.com/user-attachments/assets/b33cee47-5f36-44c7-b99c-bc027a680019" />
Dashboard:
![Screenshot 2025-02-14 at 10 53 56 AM](https://github.com/user-attachments/assets/8ab46854-f091-4164-8d86-f1ba2e5d5583)


2. It appears the goal on the graph is set to $6000, but the monthly goal is set to $7000. Is this correct?

canac commented on pages/accountLists/[accountListId].page.tsx:
```diff
@@ -106,6 +107,10 @@ export const getServerSideProps = makeGetServerSideProps(
         query: GetDashboardDocument,
         variables: {
           accountListId: query.accountListId,
+          periodBegin: DateTime.now()
+            .startOf('month')
+            .minus({ years: 1 })
+            .toISODate(),
           // TODO: implement these variables in query
```
It was loading March 2024 - February 2025 instead of February 2024 - February 2025.

canac replied:
Yes, we do want 13 months. If you look at the donations graph, it loads 13 months of donations, so we also need to 13 months of health indicator data.

canac commented:
> 1. The coaching graph looked different than the other graph's, the goal was just a straight line rather being a dot with a line.

I removed the dots from the line.

> 2. It appears the goal on the graph is set to $6000, but the monthly goal is set to $7000. Is this correct?

Yes, this is correct. It is graphing the `staffEnteredGoal`, which is $6000, while the `machineCalculatedGoal` is $7000.

@dr-bizz I also made some additional fixes and changes:

* As a coach when viewing another user's donations, clicking on the donation graph would take you to your own donations page for that month. I changed it to do nothing. You can still click on a month on your own coaching report to go to the donations page.
* For each period, if the staff entered goal is not available use the machine calculated goal instead.
* Fixed the colors of the currency bars.
* Show the goal legend in the card header when a goal is not set in preferences.

dr-bizz commented on src/components/Coaching/CoachingDetail/CoachingDetail.tsx:
```diff
@@ -148,6 +158,17 @@ export const CoachingDetail: React.FC<CoachingDetailProps> = ({
     }
   }, [sidebarDrawer]);
 
+  const handlePeriodClick = (period: DateTime) => {
+    if (accountListType === AccountListTypeEnum.Own) {
+      push({
+        pathname: `/accountLists/${accountListId}/reports/donations`,
+        query: {
+          month: period.toISODate(),
+        },
+      });
+    }
+  };
```
test

dr-bizz commented on src/components/Dashboard/DonationHistories/DonationHistories.test.tsx:
I would imagine it's quite hard to test the onPeriodClick click.

canac replied:
Unfortunately, you're right. At least not without mocking `ComposedChart`. I don't think that the effort to do that would be worth it.

dr-bizz replied:
No, it's not worth it

canac replied:
I'm just saying that since we have to mock the `ComposedChart` component, we have test coverage but no real guarantee that our click handler works with the actual chart. It's also a trivial handler that is easy to manually verify. This feels like a good candidate for the 10-20% of code that we leave without 100% test coverage.

But I went ahead and added tests for you.

dr-bizz reviewed (approved):
Looks awesome. Thank you for the improvments!

dr-bizz reviewed (commented):
Thank you for making these improvements. I reviewed it again and it looks great.

I have 1 question; It looks like when I removed the goal the machine-calulated goal isn't showing on the graph. It should be the same currency, so it should show correct? _This s probably me not understanding it correctly._

<img width="1190" alt="Screenshot 2025-02-21 at 9 38 52 AM" src="https://github.com/user-attachments/assets/6728b53f-0aa7-41b7-bcfe-af12ede9b6eb" />

canac commented:
> It looks like when I removed the goal, the machine-calculated goal isn't showing on the graph. It should be the same currency, so it should show correct?

The blue line is the staff-entered goal for each of the months in 2024. Because it has a staff-entered goal in the health indicator data it is using that and not falling back to the machine-calculated goal. The months in 2025 have no data because there is no health indicator data for them. I should extrapolate from the last known machine-calculated goal though to fill in those missing months.

dr-bizz reviewed (approved):
Excellent work on this @canac!! Thank you for fixing all the comments I've made.
